### PR TITLE
Implementing `switch` expressions: everything in `flutter/lib/src/`

### PR DESCRIPTION
**Pull request by nate-thegrate**
This PR is the 9ᵗʰ step in the journey to solve issue #136139 and make the entire Flutter repo more readable.

(previous pull requests: #139048, #139882, #141591, #142279, #142634, #142793, #143293, #143496)

I did a pass through all of `packages/flutter/lib/src/` and found an abundance of `switch` statements to improve. Whereas #143496 focused on in-depth refactoring, this PR is full of simple, straightforward changes. (I ended up making some more complicated changes in `rendering/` and will file those separately after this PR is done.)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/sliver_padding.dart [hunk 1]
```diff
@@ -103,22 +93,18 @@ abstract class RenderSliverEdgeInsetsPadding extends RenderSliver with RenderObj
   @override
   void performLayout() {
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/rendering/sliver_padding.dart:94]: This `performLayout()` function is the one spot where I did some in-depth refactoring. Originally, there was a `switch` statement with an atrocious line length:

```dart
abstract class RenderSliverEdgeInsetsPadding extends RenderSliver with RenderObjectWithChildMixin<RenderSliver> {
  void performLayout() {
    switch (applyGrowthDirectionToAxisDirection(constraints.axisDirection, constraints.growthDirection)) {
      case AxisDirection.up:
        childParentData.paintOffset = Offset(resolvedPadding!.left, calculatePaintOffset(constraints, from: resolvedPadding!.bottom + childLayoutGeometry.scrollExtent, to: resolvedPadding!.bottom + childLayoutGeometry.scrollExtent + resolvedPadding!.top));
      ...
```

After taking multiple steps to clean things up, it now looks much more reasonable:

```dart
abstract class RenderSliverEdgeInsetsPadding extends RenderSliver with RenderObjectWithChildMixin<RenderSliver> {
  void performLayout() {
    final double calculatedOffset = switch (applyGrowthDirectionToAxisDirection(constraints.axisDirection, constraints.growthDirection)) {
      AxisDirection.up => paintOffset(from: resolvedPadding.bottom + scrollExtent, to: resolvedPadding.vertical + scrollExtent),
      ...
```

File: packages/flutter/lib/src/widgets/single_child_scroll_view.dart [hunk 2]
```diff
@@ -422,53 +420,37 @@ class _RenderSingleChildViewport extends RenderBox with RenderObjectWithChildMix
...
+    };
   }
 
   @override
   double computeMinIntrinsicWidth(double height) {
-    if (child != null) {
-      return child!.getMinIntrinsicWidth(height);
-    }
-    return 0.0;
+    return child?.getMinIntrinsicWidth(height) ?? 0.0;
```
Comment (nate-thegrate) [inline packages/flutter/lib/src/widgets/single_child_scroll_view.dart:438]: This probably shouldn't be part of a pull request labeled "Implementing `switch` expressions", but it's a very simple change that makes things look nicer :)

File: packages/flutter/lib/src/material/drawer.dart [hunk 3]
```diff
@@ -644,35 +644,24 @@ class DrawerControllerState extends State<DrawerController> with SingleTickerPro
...
-      case DrawerAlignment.start:
-        return AlignmentDirectional.centerEnd;
-      case DrawerAlignment.end:
-        return AlignmentDirectional.centerStart;
-    }
+    return switch (widget.alignment) {
+      DrawerAlignment.start => AlignmentDirectional.centerEnd,
+      DrawerAlignment.end => AlignmentDirectional.centerStart,
+    };
   }
```
Comment (bernaferrari) [inline packages/flutter/lib/src/material/drawer.dart:658]: Theoretically, if you want things even shorter, I think you could even do:

`AlignmentDirectional get _drawerInnerAlignment => switch(...)`

Right?

Comment (nate-thegrate) [reply packages/flutter/lib/src/material/drawer.dart:658]: I like using that format for my own projects, but I've been using `return switch(...)` here to follow the [Flutter Style Guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#consider-using--for-short-functions-and-methods):

> only use `=>` when everything, including the function declaration, fits on a single line.

Comment (bernaferrari) [reply packages/flutter/lib/src/material/drawer.dart:658]: oh! That makes sense. Probably the rule is older than switch :P but still valid. Thanks

Comment (goderbauer) [reply packages/flutter/lib/src/material/drawer.dart:658]: One rule down in the style guide there is also this one: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use--for-inline-callbacks-that-just-return-literals-or-switch-expressions

So, changing this to what Bernardo suggested would be within the rules of the style guide.

File: packages/flutter/lib/src/cupertino/date_picker.dart [hunk 4]
```diff
@@ -414,20 +414,18 @@ class CupertinoDatePicker extends StatefulWidget {
...
-    switch (mode) {
-      case CupertinoDatePickerMode.time:
-      case CupertinoDatePickerMode.dateAndTime:
-        return _CupertinoDatePickerDateTimeState();
-      case CupertinoDatePickerMode.date:
-        return _CupertinoDatePickerDateState(dateOrder: dateOrder);
-      case CupertinoDatePickerMode.monthYear:
-        return _CupertinoDatePickerMonthYearState(dateOrder: dateOrder);
-    }
+    // ignore: no_logic_in_create_state, https://github.com/flutter/flutter/issues/70499
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/date_picker.dart:422]: This ignore made more sense at its old location. Can it be moved back? Or at the very least above the comment?

File: packages/flutter/lib/src/cupertino/route.dart [hunk 5]
```diff
@@ -975,16 +975,10 @@ class _CupertinoEdgeShadowPainter extends BoxPainter {
 
     final TextDirection? textDirection = configuration.textDirection;
     assert(textDirection != null);
-    final double start;
-    final double shadowDirection; // -1 for ltr, 1 for rtl.
```
Comment (goderbauer) [inline packages/flutter/lib/src/cupertino/route.dart:979]: This comment was lost in the refactoring, but it is kinda useful. Can you preserve it?

File: packages/flutter/lib/src/material/ink_well.dart [hunk 6]
```diff
@@ -1094,24 +1088,17 @@ class _InkResponseState extends State<_InkResponseStateWidget>
...
   void updateFocusHighlights() {
-    final bool showFocus;
-    switch (FocusManager.instance.highlightMode) {
-      case FocusHighlightMode.touch:
-        showFocus = false;
-      case FocusHighlightMode.traditional:
-        showFocus = _shouldShowFocus;
-    }
-    updateHighlight(_HighlightType.focus, value: showFocus);
+    updateHighlight(_HighlightType.focus, value: switch (FocusManager.instance.highlightMode) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/ink_well.dart:1098]: assigning the result of the switch to a variable and passing it to `value` here would IMO make this a little more readable.

File: packages/flutter/lib/src/material/drawer.dart (hunk 3 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/drawer.dart:658]: I'm glad that rule includes `switch` expressions!

I don't think it applies to class getters and methods, but I'll update `focus_traversal.dart` right now 🙂

File: packages/flutter/lib/src/cupertino/route.dart (hunk 5 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/cupertino/route.dart:979]: shadowDirection

This is one of those borderline cases as far as whether a comment is useful: in my opinion, the `switch` expression makes it easy to tell that the value of `shadowDirection` is `1` for `rtl` and `-1` for `ltr`.

File: packages/flutter/lib/src/cupertino/date_picker.dart (hunk 4 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/cupertino/date_picker.dart:418]: Unfortunately, it needs to be placed right above the `switch` expression in order to suppress the warning.

But I was able to reposition the other comments; hopefully it's easier to follow now.

File: packages/flutter/lib/src/material/ink_well.dart (hunk 6 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/ink_well.dart]: done!

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/services/raw_keyboard_android.dart [hunk 7]
```diff
@@ -216,40 +216,27 @@ class RawKeyEventDataAndroid extends RawKeyEventData {
...
-      case KeyboardSide.all:
-        return metaState & leftMask != 0 && metaState & rightMask != 0;
-      case KeyboardSide.left:
-        return metaState & leftMask != 0;
-      case KeyboardSide.right:
-        return metaState & rightMask != 0;
-    }
+    return switch (side) {
+      KeyboardSide.any   => true,
+      KeyboardSide.all   => metaState & leftMask != 0 && metaState & rightMask != 0,
```
Comment (loic-sharma) [inline packages/flutter/lib/src/services/raw_keyboard_android.dart:221]: Thoughts on this? I don't feel strongly:

```suggestion
      KeyboardSide.all   => (metaState & leftMask != 0) && (metaState & rightMask != 0),
```

File: packages/flutter/lib/src/services/raw_keyboard_fuchsia.dart [hunk 8]
```diff
@@ -99,16 +99,12 @@ class RawKeyEventDataFuchsia extends RawKeyEventData {
...
-      case KeyboardSide.all:
-        return modifiers & leftMask != 0 && modifiers & rightMask != 0;
-      case KeyboardSide.left:
-        return modifiers & leftMask != 0;
-      case KeyboardSide.right:
-        return modifiers & rightMask != 0;
-    }
+    return switch (side) {
+      KeyboardSide.any   => true,
+      KeyboardSide.all   => modifiers & leftMask != 0 && modifiers & rightMask != 0,
```
Comment (loic-sharma) [inline packages/flutter/lib/src/services/raw_keyboard_fuchsia.dart:104]: If you change the case above, consider updating this as well:

```suggestion
      KeyboardSide.all   => (modifiers & leftMask != 0) && (modifiers & rightMask != 0),
```

File: packages/flutter/lib/src/services/raw_keyboard_android.dart (hunk 7 above)
Comment (nate-thegrate) [reply packages/flutter/lib/src/services/raw_keyboard_android.dart]: Works for me!

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 9]
```diff
@@ -548,16 +548,12 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
...
-      case AxisDirection.left:
-        return Offset(-position.pixels, 0);
-      case AxisDirection.right:
-        return Offset(position.pixels, 0);
-    }
+    return switch (axisDirection) {
+      AxisDirection.down  => Offset(0,  position.pixels),
+      AxisDirection.up    => Offset(0, -position.pixels),
+      AxisDirection.left  => Offset(-position.pixels, 0),
+      AxisDirection.right => Offset( position.pixels, 0),
```
Comment (loic-sharma) [inline packages/flutter/lib/src/widgets/scrollable.dart:555]: This looks a bit funky to me, consider instead:

```suggestion
      AxisDirection.right => Offset(position.pixels,  0),
```

Or:

```suggestion
      AxisDirection.right => Offset(position.pixels, 0),
```

File: packages/flutter/lib/src/widgets/scrollable.dart [hunk 10]
```diff
@@ -1491,16 +1487,12 @@ class _ScrollableSelectionContainerDelegate extends MultiSelectableSelectionCont
...
-    case AxisDirection.left:
-      return Offset(-scrollableState.position.pixels, 0);
-    case AxisDirection.right:
-      return Offset(scrollableState.position.pixels, 0);
-  }
+  return switch (scrollableState.axisDirection) {
+    AxisDirection.down  => Offset(0,  scrollableState.position.pixels),
+    AxisDirection.up    => Offset(0, -scrollableState.position.pixels),
+    AxisDirection.left  => Offset(-scrollableState.position.pixels, 0),
+    AxisDirection.right => Offset( scrollableState.position.pixels, 0),
```
Comment (loic-sharma) [inline packages/flutter/lib/src/widgets/scrollable.dart:1494]: Same comment here as above: https://github.com/flutter/flutter/pull/143634/files#r1496378734

**Review by loic-sharma — APPROVED** [review]
Awesome improvements, thanks for the contribution!

**Pull request merged**